### DISCUSSION TITLE
[ Fundamentos do C# - Aula .NET Core] Diferenças entre o .NET Framework e .NET Core

### DISCUSSION THREAD

**Discussion by lvsantos** [balta-io/forum#392] [Tira Dúvidas] [answered]
Nesta aula é falado sobre a chegada do .NET Core e suas mudanças em relação ao .NET Framework. Eu fiquei com as seguintes dúvidas com relação a essa reescrita.

1. A linguagem C# foi reescrita para o .NET Core ou foram feitas apenas melhorias para otimizar a sua compilação e legibilidade?
2. O CLI (Common Language Infrastructure) e o CLR(Common Language Runtime) do .NET Core foram criados do zero?
- Se sim, quer dizer que a IL do .NET Framework é diferente da IL do .NET Core, assim como o JIT?

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @lvsantos, como vai?

A linguagem C# foi reescrita para o .NET Core ou foram feitas apenas melhorias para otimizar a sua compilação e legibilidade?
- O C# nasceu junto com o .NET


O CLI (Common Language Infrastructure) e o CLR(Common Language Runtime) do .NET Core foram criados do zero?
- Sim, o CLR foi criado do zero

Se sim, quer dizer que a IL do .NET Framework é diferente da IL do .NET Core, assim como o JIT?
- IL é o Intermediate Language (linguagem intermediária), esta é a linguagem que o C# é convertido antes de ser compilado.
- A IL é a mesma para ambos (Ela sempre recebe otimizações)
- O compilador, no Core passou a ser o Roslyn, que é mais eficiente e sempre recebe otimizações.

**Reply by lvsantos** [reply]
Ok @andrebaltieri ficou mais claro agora, mas ainda ficaram mais algumas dúvidas hehe. 
1) Em relação ao CLR, eu posso dizer que ele atua apenas após a geração da IL, ou ele também está presente durante a compilação do código C#?
2) A primeira compilação que um código C# passa eu posso chamá-la de AOT (Ahead of time) e a segunda de JIT (Just in time) ?

Eu fiz o seguinte desenho para condensar melhor as ideias:

![image](https://github.com/balta-io/forum/assets/34079199/f45cf1ee-1a79-45ca-869e-d70ae06b47a6)

Em relação as diferenças entre o .NET Framework e o .NET Core ficaram as seguintes dúvidas:

3) A partir da versão 5 tudo foi unificado, porém existiu uma época em que a linguagem C# era usada tanto para o .NET Framework quanto para o Core, por exemplo a versão 7.3 do C#. Então para essa versão, eu posso afirmar que o processo de compilação para a linguagem IL era diferente para cada framework, porém eles geravam a mesma IL?
4) Como o CLR de ambas eram diferentes, a forma como o JIT atuava em cada uma também era diferente, o que provavelmente era uma vantagem do .NET Core por ter tido seu CLR criado do zero?

Novamente fiz um desenho para condensar melhor as ideias:

![image](https://github.com/balta-io/forum/assets/34079199/0aa33c9e-3192-4f6c-8ff0-928b25827490)

**Reply by andrebaltieri** [reply]
Em relação ao CLR, eu posso dizer que ele atua apenas após a geração da IL, ou ele também está presente durante a compilação do código C#?
- Isto, o Runtime executa o código de forma gerenciada

A primeira compilação que um código C# passa eu posso chamá-la de AOT (Ahead of time) e a segunda de JIT (Just in time) ?
- Humm não, a primeira "compilação" age mais como uma tradução do C# para linguagem intermediária.

A partir da versão 5 tudo foi unificado, porém existiu uma época em que a linguagem C# era usada tanto para o .NET Framework quanto para o Core, por exemplo a versão 7.3 do C#. Então para essa versão, eu posso afirmar que o processo de compilação para a linguagem IL era diferente para cada framework, porém eles geravam a mesma IL?
- O C# e o .NET se complementam mas andam separados, o .NET é escrito em C#. Então sim, este processo que gera o IL é constantemente alterado, por isso as vezes temos melhorias de performance que não necessitam de nenhuma alteração no código por exemplo.

Como o CLR de ambas eram diferentes, a forma como o JIT atuava em cada uma também era diferente, o que provavelmente era uma vantagem do .NET Core por ter tido seu CLR criado do zero?
- O CLR não era diferente, ele está relacionado a versão do .NET instalada. Se você está usando o .NET 7 em um projeto ele vai usar o C# 11 e o CLR daquela versão. Mudando a IL, o JIT atua de forma diferente sim, é o que resulta nas otimizações que temos do .NET.

Ótimas perguntas, se tiver mais, manda ai!
